### PR TITLE
add imx8m (imx-cortexa53) support

### PR DESCRIPTION
**Pull request by Gateworks**
Add imx8m support:
- add imx-cortexa53 subtarget support
- add support for Gateworks Venice (imx8mm/imx8mn/imx8mp) boards
- add some backports for imx8mp PCIe

kernel+dtb has been tested on a variety of imx8m{m,n,p}-venice* boards as well as the imx8mm-evk.

compressed disk images produced by the 2nd patch that adds support for Gateworks Venice has been tested on imx8m{m,n,p}-venice* boards.

Hoping for reviews from @pepe2k and @ynezz

### CODE REVIEW

**Review by ehem — APPROVED** [review]
Both commits look reasonable to me.  You made indeed need to wait for @pepe2k and/or @ynezz though.

File: package/kernel/linux/modules/video.mk [hunk 1]
```diff
@@ -1142,7 +1142,7 @@ $(eval $(call KernelPackage,video-dma))
 
 define KernelPackage/video-coda
   TITLE:=i.MX VPU support
-  DEPENDS:=@(TARGET_imx&&!TARGET_imx_cortexa7) +kmod-video-mem2mem +kmod-video-dma
+  DEPENDS:=@(TARGET_imx&&TARGET_imx_cortexa9) +kmod-video-mem2mem +kmod-video-dma
```
Comment (Leo-PL) [inline package/kernel/linux/modules/video.mk:1145]: Why this switch to cortexa9 here? This probably needs to go to a separate commit

**Comment by xback** [comment]
I can test this series on a fully stacked venice board somewhere next week.

File: package/kernel/linux/modules/video.mk (hunk 1 above)
Comment (Gateworks) [reply package/kernel/linux/modules/video.mk:1145]: I just inverted the logic because the modules are valid for coretexa9 but not valid for cortexa7 and cortexa53 (or at least not yet valid for cortexa53)

Comment (Gateworks) [reply package/kernel/linux/modules/video.mk:1145]: in the case of cortexa53, or at least imx8mm the vpu is hantro and not coda

Comment (Leo-PL) [reply package/kernel/linux/modules/video.mk:1145]: Okay, I missed that negation.

**Comment by xback** [comment]
@Gateworks 
As promised, I did some testing on my venice board (gw7200-00).
Results are a bit mixed.


When I flashed the OpenWRT img according to instructions mentioned in the commit, the board was broken after flashing.
It refused to print anything (not even u-boot SPL prints)

```
U-Boot SPL 2021.07-g21a6c5d7d9 (Apr 05 2022 - 21:01:11 +0000)
GSCv3   : v61 0x1d6f RST:VIN Thermal protection:disabled 
RTC     : 1970-01-01   0:00:00 UTC
Model   : GW7201-01-DF
Serial  : 902937
MFGDate : 08-03-2022
PMIC    : MP5416
DRAM    : LPDDR4 4 GiB
WDT:   Started with servicing (60s timeout)
Trying to boot from MMC1
DTB     : imx8mm-venice-gw72xx-0x
NOTICE:  BL31: v2.4(release):f884ad7b0
NOTICE:  BL31: Built : 19:58:46, Apr  5 2022


U-Boot 2021.07-g21a6c5d7d9 (Apr 05 2022 - 21:01:11 +0000), Build: jenkins-venice-bsp-107

CPU:   Freescale i.MX8MMQ rev1.0 1600 MHz (running at 1200 MHz)
CPU:   Industrial temperature grade (-40C to 105C) at 28C
Reset cause: POR
Model: Gateworks Venice GW72xx-0x i.MX8MM Development Kit
DRAM:  4 GiB
WDT:   Started with servicing (60s timeout)
MMC:   FSL_SDHC: 1, FSL_SDHC: 2
Loading Environment from MMC... OK
In:    serial
Out:   serial
Err:   serial
Net:   GPY111 eth0: ethernet@30be0000 [PRIME]
GSC     : boot watchdog disabled
Thermal protection:enabled at 96C 
Hit any key to stop autoboot:  0 
u-boot=> tftpboot $loadaddr openwrt-imx-cortexa53-gateworks_venice-squashfs-img.gz && gzwrite mmc $dev $loadaddr $filesize
Using ethernet@30be0000 device
TFTP from server 192.168.1.146; our IP address is 192.168.1.1
Filename 'openwrt-imx-cortexa53-gateworks_venice-squashfs-img.gz'.
Load address: 0x48200000
Loading: #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         #################################################################
         ############################################
         5.8 MiB/s
done
Bytes transferred = 10176551 (9b4827 hex)

142868480/335675392
        uncompressed 142868480 of 335675392
        crcs == 0x00000000/0x0a8f5a2d
u-boot=> reset
resetting ...


* BOOT STOPS - broken on re-powering *
```

Luckely, I have a few Gateworks jtag flash tools, and used them to flash the Gateworks image (u-boot, SPL, ATF, ...) and GSC fw to the latest versions.

After doing this, The board now booted after flashing OpenWRT through u-boot.
but .. It always halts after USB enumeration.
After 60 seconds, the WDT kicks in and reboots the board.


```
U-Boot SPL 2023.04-g7cf6733072 (Jan 24 2024 - 20:59:58 +0000)
GSCv3   : v64 0x4920 RST:CPU Thermal protection:enabled at 96C 
RTC     : 1970-01-01   0:05:43 UTC
Model   : GW7201-01-DF
Serial  : 902937
MFGDate : 08-03-2022
PMIC    : MP5416 (IMX8MM)
DRAM    : LPDDR4 4 GiB 3000MT/s 1500MHz
SEC0:  RNG instantiated
WDT:   Started watchdog@30280000 with servicing every 1000ms (60s timeout)
Trying to boot from eMMC
DTB     : imx8mm-venice-gw72xx-0x
NOTICE:  BL31: v2.4(release):f884ad7b0
NOTICE:  BL31: Built : 19:44:56, Jan 24 2024


U-Boot 2023.04-g7cf6733072 (Jan 24 2024 - 20:59:58 +0000), Build: jenkins-venice-bsp-170

CPU:   Freescale i.MX8MMQ rev1.0 1600 MHz (running at 1200 MHz)
CPU:   Industrial temperature grade (-40C to 105C) at 49C
Reset cause: POR
Model: Gateworks Venice GW72xx-0x i.MX8MM Development Kit
DRAM:  4 GiB
Core:  198 devices, 27 uclasses, devicetree: separate
WDT:   Started watchdog@30280000 with servicing every 1000ms (60s timeout)
MMC:   FSL_SDHC: 1, FSL_SDHC: 2
Loading Environment from MMC... OK
In:    serial@30890000
Out:   serial@30890000
Err:   serial@30890000
SEC0:  RNG instantiated
Net:   GPY111 eth0: ethernet@30be0000 [PRIME]
GSC     : boot watchdog disabled
Thermal protection:enabled at 96C 
Hit any key to stop autoboot:  0 
MMC: no card present
switch to partitions #0, OK
mmc2(part 0) is current device
Scanning mmc 2:1...
Found U-Boot script /boot.scr
2559 bytes read in 5 ms (499 KiB/s)
## Executing script at 50280000
Gateworks Venice OpenWrt Boot script v1.0
loading DTB...
Loading imx8mm-venice-gw72xx-0x.dtb...
31466 bytes read in 2 ms (15 MiB/s)
loading kernel...
12437512 bytes read in 43 ms (275.8 MiB/s)
## Flattened Device Tree blob at 50200000
   Booting using the fdt blob at 0x50200000
Working FDT set to 50200000
   Loading Device Tree to 00000000fceeb000, end 00000000fcef5ae9 ... OK
Working FDT set to fceeb000

Starting kernel ...

[    0.000000] Booting Linux on physical CPU 0x0000000000 [0x410fd034]
[    0.000000] Linux version 5.15.147 (koen@roemer) (aarch64-openwrt-linux-musl-gcc (OpenWrt GCC 12.3.0 r24862-068f685dec) 12.3.0, GNU ld (GNU Binutils) 2.40.0) #0 SMP Fri Jan 19 01:50:16 2024
[    0.000000] Machine model: Gateworks Venice GW72xx-0x i.MX8MM Development Kit
[    0.000000] earlycon: ec_imx6q0 at MMIO 0x0000000030890000 (options '115200')
[    0.000000] printk: bootconsole [ec_imx6q0] enabled
[    0.000000] Zone ranges:
[    0.000000]   DMA      [mem 0x0000000040000000-0x00000000ffffffff]
[    0.000000]   DMA32    empty
[    0.000000]   Normal   [mem 0x0000000100000000-0x000000013fffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000040000000-0x000000013fffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000040000000-0x000000013fffffff]
[    0.000000] psci: probing for conduit method from DT.
[    0.000000] psci: PSCIv1.1 detected in firmware.
[    0.000000] psci: Using standard PSCI v0.2 function IDs
[    0.000000] psci: MIGRATE_INFO_TYPE not supported.
[    0.000000] psci: SMC Calling Convention v1.2
[    0.000000] percpu: Embedded 16 pages/cpu s25304 r8192 d32040 u65536
[    0.000000] Detected VIPT I-cache on CPU0
[    0.000000] CPU features: detected: GIC system register CPU interface
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 1032192
[    0.000000] Kernel command line: net.ifnames=0 console=ttymxc1,115200 earlycon=ec_imx6q,0x30890000,115200 root=PARTUUID=175d09a7-02 rootfstype=squashfs,ext4,f2fs rootwait
[    0.000000] Dentry cache hash table entries: 524288 (order: 10, 4194304 bytes, linear)
[    0.000000] Inode-cache hash table entries: 262144 (order: 9, 2097152 bytes, linear)
[    0.000000] mem auto-init: stack:off, heap alloc:off, heap free:off
[    0.000000] software IO TLB: mapped [mem 0x00000000f8eeb000-0x00000000fceeb000] (64MB)
[    0.000000] Memory: 4035176K/4194304K available (8384K kernel code, 946K rwdata, 2224K rodata, 448K init, 291K bss, 159128K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=64, Order=0-3, MinObjects=0, CPUs=4, Nodes=1
[    0.000000] rcu: Hierarchical RCU implementation.
[    0.000000]  Tracing variant of Tasks RCU enabled.
[    0.000000] rcu: RCU calculated value of scheduler-enlistment delay is 10 jiffies.
[    0.000000] NR_IRQS: 64, nr_irqs: 64, preallocated irqs: 0
[    0.000000] GICv3: GIC: Using split EOI/Deactivate mode
[    0.000000] GICv3: 128 SPIs implemented
[    0.000000] GICv3: 0 Extended SPIs implemented
[    0.000000] GICv3: Distributor has no Range Selector support
[    0.000000] Root IRQ handler: gic_handle_irq
[    0.000000] GICv3: 16 PPIs implemented
[    0.000000] GICv3: CPU0: found redistributor 0 region 0:0x0000000038880000
[    0.000000] ITS: No ITS available, not enabling LPIs
[    0.000000] arch_timer: cp15 timer(s) running at 8.00MHz (phys).
[    0.000000] clocksource: arch_sys_counter: mask: 0xffffffffffffff max_cycles: 0x1d854df40, max_idle_ns: 440795202120 ns
[    0.000000] sched_clock: 56 bits at 8MHz, resolution 125ns, wraps every 2199023255500ns
[    0.008309] Console: colour dummy device 80x25
[    0.012533] Calibrating delay loop (skipped), value calculated using timer frequency.. 16.00 BogoMIPS (lpj=80000)
[    0.022841] pid_max: default: 32768 minimum: 301
[    0.027830] Mount-cache hash table entries: 8192 (order: 4, 65536 bytes, linear)
[    0.034925] Mountpoint-cache hash table entries: 8192 (order: 4, 65536 bytes, linear)
[    0.044241] rcu: Hierarchical SRCU implementation.
[    0.048057] smp: Bringing up secondary CPUs ...
[    0.052506] Detected VIPT I-cache on CPU1
[    0.052534] GICv3: CPU1: found redistributor 1 region 0:0x00000000388a0000
[    0.052569] CPU1: Booted secondary processor 0x0000000001 [0x410fd034]
[    0.052942] Detected VIPT I-cache on CPU2
[    0.052961] GICv3: CPU2: found redistributor 2 region 0:0x00000000388c0000
[    0.052981] CPU2: Booted secondary processor 0x0000000002 [0x410fd034]
[    0.053331] Detected VIPT I-cache on CPU3
[    0.053350] GICv3: CPU3: found redistributor 3 region 0:0x00000000388e0000
[    0.053369] CPU3: Booted secondary processor 0x0000000003 [0x410fd034]
[    0.053415] smp: Brought up 1 node, 4 CPUs
[    0.108764] SMP: Total of 4 processors activated.
[    0.113486] CPU features: detected: 32-bit EL0 Support
[    0.118648] CPU features: detected: CRC32 instructions
[    0.123841] CPU features: emulated: Privileged Access Never (PAN) using TTBR0_EL1 switching
[    0.132211] CPU: All CPU(s) started at EL2
[    0.136339] alternatives: patching kernel code
[    0.146465] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.153503] futex hash table entries: 1024 (order: 4, 65536 bytes, linear)
[    0.160625] pinctrl core: initialized pinctrl subsystem
[    0.166703] NET: Registered PF_NETLINK/PF_ROUTE protocol family
[    0.172024] DMA: preallocated 512 KiB GFP_KERNEL pool for atomic allocations
[    0.178756] DMA: preallocated 512 KiB GFP_KERNEL|GFP_DMA pool for atomic allocations
[    0.186548] DMA: preallocated 512 KiB GFP_KERNEL|GFP_DMA32 pool for atomic allocations
[    0.194825] thermal_sys: Registered thermal governor 'step_wise'
[    0.195038] ASID allocator initialised with 65536 entries
[    0.211628] imx8mm-pinctrl 30330000.pinctrl: initialized IMX pinctrl driver
[    0.228765] cryptd: max_cpu_qlen set to 1000
[    0.233496] SCSI subsystem initialized
[    0.234593] usbcore: registered new interface driver usbfs
[    0.239931] usbcore: registered new interface driver hub
[    0.245297] usbcore: registered new device driver usb
[    0.250877] pps_core: LinuxPPS API ver. 1 registered
[    0.255330] pps_core: Software ver. 5.3.6 - Copyright 2005-2007 Rodolfo Giometti <giometti@linux.it>
[    0.264507] PTP clock support registered
[    0.269335] clocksource: Switched to clocksource arch_sys_counter
[    0.277827] NET: Registered PF_INET protocol family
[    0.280042] IP idents hash table entries: 65536 (order: 7, 524288 bytes, linear)
[    0.288915] tcp_listen_portaddr_hash hash table entries: 2048 (order: 3, 32768 bytes, linear)
[    0.295915] Table-perturb hash table entries: 65536 (order: 6, 262144 bytes, linear)
[    0.303672] TCP established hash table entries: 32768 (order: 6, 262144 bytes, linear)
[    0.311830] TCP bind hash table entries: 32768 (order: 7, 524288 bytes, linear)
[    0.319453] TCP: Hash tables configured (established 32768 bind 32768)
[    0.325643] UDP hash table entries: 2048 (order: 4, 65536 bytes, linear)
[    0.332350] UDP-Lite hash table entries: 2048 (order: 4, 65536 bytes, linear)
[    0.339690] NET: Registered PF_UNIX/PF_LOCAL protocol family
[    0.345154] PCI: CLS 0 bytes, default 64
[    0.352210] workingset: timestamp_bits=46 max_order=20 bucket_order=0
[    0.359780] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.411979] mxs-dma 33000000.dma-controller: initialized
[    0.414998] imx-sdma 302c0000.dma-controller: loaded firmware 4.5
[    0.420739] SoC: i.MX8MM revision 1.0
[    0.424851] Serial: 8250/16550 driver, 2 ports, IRQ sharing disabled
[    0.431323] 30860000.serial: ttymxc0 at MMIO 0x30860000 (irq = 34, base_baud = 1500000) is a IMX
[    0.439991] 30880000.serial: ttymxc2 at MMIO 0x30880000 (irq = 35, base_baud = 1500000) is a IMX
[    0.448678] 30890000.serial: ttymxc1 at MMIO 0x30890000 (irq = 36, base_baud = 1500000) is a IMX
[    0.457196] printk: console [ttymxc1] enabled
[    0.457196] printk: console [ttymxc1] enabled
[    0.465864] printk: bootconsole [ec_imx6q0] disabled
[    0.465864] printk: bootconsole [ec_imx6q0] disabled
[    0.476195] 30a60000.serial: ttymxc3 at MMIO 0x30a60000 (irq = 41, base_baud = 1500000) is a IMX
[    0.488072] imx-sdma 302b0000.dma-controller: loaded firmware 4.5
[    0.494378] imx-sdma 30bd0000.dma-controller: loaded firmware 4.5
[    0.502939] loop: module loaded
[    0.508226] pps pps0: new PPS source ptp0
[    0.514239] fec 30be0000.ethernet eth0: registered PHC device 0
[    0.520722] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    0.527300] usbcore: registered new interface driver usb-storage
[    0.534032] i2c_dev: i2c /dev entries driver
[    0.539501] sdhci: Secure Digital Host Controller Interface driver
[    0.545692] sdhci: Copyright(c) Pierre Ossman
[    0.550082] sdhci-pltfm: SDHCI platform and OF driver helper
[    0.556618] caam 30900000.crypto: device ID = 0x0a16040100000000 (Era 9)
[    0.563349] caam 30900000.crypto: job rings = 3, qi = 0
[    0.569906] caam_jr 30901000.jr: failed to flush job ring 0
[    0.575492] caam_jr: probe of 30901000.jr failed with error -5
[    0.584735] caam algorithms registered in /proc/crypto
[    0.588147] mmc2: SDHCI controller on 30b60000.mmc [30b60000.mmc] using ADMA
[    0.590732] caam 30900000.crypto: caam pkc algorithms registered in /proc/crypto
[    0.604364] caam 30900000.crypto: registering rng-caam
[    0.611612] NET: Registered PF_INET6 protocol family
[    0.617367] Segment Routing with IPv6
[    0.621077] In-situ OAM (IOAM) with IPv6
[    0.625040] NET: Registered PF_PACKET protocol family
[    0.630222] 8021q: 802.1Q VLAN Support v1.8
[    0.634678] Key type .fscrypt registered
[    0.638602] Key type fscrypt-provisioning registered
[    0.644771] Key type encrypted registered
[    0.665900] usb_phy_generic usbphynop1: supply vcc not found, using dummy regulator
[    0.668604] mmc2: new HS400 Enhanced strobe MMC card at address 0001
[    0.673677] usb_phy_generic usbphynop1: dummy supplies not allowed for exclusive requests
[    0.688295] usb_phy_generic usbphynop2: supply vcc not found, using dummy regulator
[    0.689675] mmcblk2: mmc2:0001 DA6064 58.2 GiB 
[    0.696075] usb_phy_generic usbphynop2: dummy supplies not allowed for exclusive requests
[    0.702237]  mmcblk2: p1 p2
[    0.712540] mmcblk2boot0: mmc2:0001 DA6064 4.00 MiB 
[    0.718055] mmcblk2boot1: mmc2:0001 DA6064 4.00 MiB 
[    0.723356] mmcblk2rpmb: mmc2:0001 DA6064 4.00 MiB, chardev (249:0)
[    0.731184] i2c i2c-0: IMX I2C adapter registered
[    0.736799] i2c i2c-1: IMX I2C adapter registered
[    0.742160] i2c i2c-2: IMX I2C adapter registered
[    0.754288] ci_hdrc ci_hdrc.1: EHCI Host Controller
[    0.759194] ci_hdrc ci_hdrc.1: new USB bus registered, assigned bus number 1
[    0.789338] ci_hdrc ci_hdrc.1: USB 2.0 started, EHCI 1.00
[    0.790069] random: crng init done
[    0.795109] hub 1-0:1.0: USB hub found
[    0.801916] hub 1-0:1.0: 1 port detected

*freeze*  WDT kicks in after 60 sec
```

Once in a while, it prints a little bit more:


```
[    0.670182] usb_phy_generic usbphynop1: dummy supplies not allowed for exclusive requests
[    0.678532] usb_phy_generic usbphynop2: supply vcc not found, using dummy regulator
[    0.686281] usb_phy_generic usbphynop2: dummy supplies not allowed for exclusive requests
[    0.704055] mmc2: new HS400 Enhanced strobe MMC card at address 0001
[    0.710905] mmcblk2: mmc2:0001 DA6064 58.2 GiB 
[    0.716012]  mmcblk2: p1 p2
[    0.719259] mmcblk2boot0: mmc2:0001 DA6064 4.00 MiB 
[    0.721868] i2c i2c-0: IMX I2C adapter registered
[    0.724669] mmcblk2boot1: mmc2:0001 DA6064 4.00 MiB 
[    0.729888] i2c i2c-1: IMX I2C adapter registered
[    0.734235] mmcblk2rpmb: mmc2:0001 DA6064 4.00 MiB, chardev (249:0)
[    0.739275] i2c i2c-2: IMX I2C adapter registered
[    0.756925] ci_hdrc ci_hdrc.1: EHCI Host Controller
[    0.761893] ci_hdrc ci_hdrc.1: new USB bus registered, assigned bus number 1
[    0.786697] random: crng init done
[    0.789293] ci_hdrc ci_hdrc.1: USB 2.0 started, EHCI 1.00
[    0.795914] hub 1-0:1.0: USB hub found
[    0.799707] hub 1-0:1.0: 1 port detected
[   60.859197] rcu: INFO: rcu_sched detected stalls on CPUs/tasks:
[   60.865119] rcu:     0-...!: (0 ticks this GP) idle=7d5/1/0x4000000000000000 softirq=49/49 fqs=0 
[   60.873738]  (detected by 3, t=6006 jiffies, g=-1131, q=97)
[   60.879312] Task dump for CPU 0:
[   60.882540] task:kworker/u8:0    state:R  running task     stack:    0 pid:    9 ppid:     2 flags:0x0000000a
[   60.892463] Workqueue: events_unbound deferred_probe_work_func
[   60.898306] Call trace:
[   60.900751]  __switch_to+0xd4/0x114
[   60.904245]  0xc00
[   60.906259] rcu: rcu_sched kthread timer wakeup didn't happen for 6005 jiffies! g-1131 f0x0 RCU_GP_WAIT_FQS(5) ->state=0x402
[   60.917474] rcu:     Possible timer handling issue on cpu=0 timer-softirq=21
[   60.924264] rcu: rcu_sched kthread starved for 6011 jiffies! g-1131 f0x0 RCU_GP_WAIT_FQS(5) ->state=0x402 ->cpu=0
[   60.934524] rcu:     Unless rcu_sched kthread gets sufficient CPU time, OOM is now expected behavior.
[   60.943483] rcu: RCU grace-period kthread stack dump:
[   60.948532] task:rcu_sched       state:I stack:    0 pid:   13 ppid:     2 flags:0x00000008
[   60.956886] Call trace:
[   60.959330]  __switch_to+0xd4/0x114
[   60.962821]  __schedule+0x238/0x580
[   60.966314]  schedule+0x58/0xc0
[   60.969460]  schedule_timeout+0x6c/0xc4
[   60.973299]  rcu_gp_fqs_loop+0xdc/0x360
[   60.977141]  rcu_gp_kthread+0xe4/0x11c
[   60.980894]  kthread+0x11c/0x130
[   60.984128]  ret_from_fork+0x10/0x20
[   60.987703] rcu: Stack dump where RCU GP kthread last ran:
[   60.993186] Task dump for CPU 0:
[   60.996413] task:kworker/u8:0    state:R  running task     stack:    0 pid:    9 ppid:     2 flags:0x0000000a
[   61.006330] Workqueue: events_unbound deferred_probe_work_func
[   61.012165] Call trace:
```

Please let me know if I'm doing something wrong here.
I also tested the precompiled OpenWRT image from the Gateworks site (jtag based) and the results are identical.

Carrier PCB: 0221097-03
SOM PCB: 02210210-05 
SOM Serial: 902937

**Comment by Gateworks** [comment]
@xback,

Thanks for testing. We switched our boot firmware from the emmc user hardware partition to the boot0 hardware partition recently to provide for a more physical isolation between the OS and the boot firmware as well as to allow the OS firmware to support imx8mm/imx8mn/imx8mp as only the boot firmware changes between those. The result is that the compressed disk image does not contain the boot firmware. In your case your board still had boot firmware on the user hardware partition so when you updated via the compressed disk image you bricked the board because you wrote over the boot firmware. When you used JTAG to program on new boot firmware you resolved that as well as got the boot firmware onto boot0 so that you won't encounter that anymore. I should update the 2nd commit to explain this.

The hang at usb is due to missing blk-cntl patches which are in the 6.1 kernel. You should be fine if you use the 6.1 testing kernel.

**Comment by xback** [comment]
@Gateworks 
Thanks for the details!

>  The result is that the compressed disk image does not contain the boot firmware. In your case your board still had boot firmware on the user hardware partition so when you updated via the compressed disk image you bricked the board because you wrote over the boot firmware

oh ..
I think this needs a big fat warning in the commit at least.
Which minimum bootloader version is required to avoid this issue? (if known, please mention it in the commit also)
People who were early adopters, without proper jtag hardware, are in big trouble otherwise.


> The hang at usb is due to missing blk-cntl patches which are in the 6.1 kernel. You should be fine if you use the 6.1 testing kernel.

So this means this currently is not usable until the target is at kernel >= 6.1 ?
Out of curiousity: does this mean the venice openwrt 5.15 images, hosted on the Gateworks website, are .. useless? :-)

**Comment by xback** [comment]
> > @Gateworks Thanks for the details!
> > > The result is that the compressed disk image does not contain the boot firmware. In your case your board still had boot firmware on the user hardware partition so when you updated via the compressed disk image you bricked the board because you wrote over the boot firmware
> > 
> > 
> > oh .. I think this needs a big fat warning in the commit at least. Which minimum bootloader version is required to avoid this issue? (if known, please mention it in the commit also) People who were early adopters, without proper jtag hardware, are in big trouble otherwise.
> 
> agreed - I have updated via a force push to add additional info to the commit log

Great! Thank you
> 
> It's less about a specific version of bootloader (although there is a change in the bootloader to where the env is stored) and more about the process of installing boot firmware to boot0 (different emmc device, different offset, changing the partconf register specifying which hardware partition is active on powerup). The JTAG programming operation handles it for you as it embedds in the JTAG image the partition, partconf register setting, and offset.
> 
> > > The hang at usb is due to missing blk-cntl patches which are in the 6.1 kernel. You should be fine if you use the 6.1 testing kernel.
> > 
> > 
> > So this means this currently is not usable until the target is at kernel >= 6.1 ? Out of curiousity: does this mean the venice openwrt 5.15 images, hosted on the Gateworks website, are .. useless? :-)
> 
> I guess it depends on what is considered usable. Some imx8m boards will boot fine depending on if/how they configure USB but PCIe isn't supported. In general:
> 
>     * imx8mm pcie was added in 5.18
> 
>     * imx8mp pcie was added in 6.1 but it needs some phy support that went into 6.2 (I have backports for it queued up)
> 
>     * imx8mp eqos MAC also needs some backports for 6.1 (I have those queue up also)
> 
>     * imx8mm and imx8mp power domains received some necessary fixes for USB power domain and PCI power domain sometime between 5.15 and 6.1 (which is what's causing the hang you see)
> 

Ok, thanks for explaining this in detail.
Reading the above list, I still would wait until 6.1 is the default on i.mx, no?

> 
> The Venice openwrt 5.15 images you refer to are from an openwrt branch where I backports a ton of patches to fix all of the above in 5.15 so they work fine but only support imx8mm/imx8mn and not imx8mp.

**Comment by Gateworks** [comment]
> Ok, thanks for explaining this in detail. Reading the above list, I still would wait until 6.1 is the default on i.mx, no?

I don't think its necessary to wait. I can add a comment in the first patch if necessary stating what works and what doesn't per 5.15 and 6.1 if needed. But again, there are several imx8mp features that don't work in 6.1 until I push some additional backport patches as well and I'm sure as more testing gets done there will be more patches on top of that. There's no point in waiting until everything is perfect as nothing is perfect and everything is a moving target :)

This is a stepping stone. As long as its not breaking the existing subtargets or doing something fundamentally wrong I would think it should be merged.

**Comment by xback** [comment]
@Gateworks 

>  I can add a comment in the first patch if necessary stating what works and what doesn't per 5.15 and 6.1 if needed
Please do!

>  But again, there are several imx8mp features that don't work in 6.1 until I push some additional backport patches as well

I'm not worrying about 6.1, the backports or the missing features.
My point here is that the current kernel is still 5.15, in which case my board didn't even boot properly :-)
Images using 5.15 will get build automatically when this gets merged as-is.

I'm not trying to be annoying here .. but I personally don't see the advantage of having images with 5.15 if all important bits are
>=5.18 and none of them get backported to 5.15.

I tested the current state like a typical user would do and it didn't boot :-)
I **do** fully agree that it's a totally different ball game when 6.1 is the default kernel on this target.

Maybe another idea would be to declare it "broken" to avoid producing images in the meantime?

@pepe2k @Ansuel 
Do you guys have an opinion on this one?

**Comment by xback** [comment]
@Gateworks I'm conducting lots of stress testing on various imx boards currently.
I haven't seen any issues so far so I guess 6.1 will be the default pretty quickly

**Comment by xback** [comment]
@Gateworks 
imx is now on 6.1 by default.
Could you kindly rebase this series to the latest master state?

Thanks!

**Comment by Gateworks** [comment]
> @Gateworks imx is now on 6.1 by default. Could you kindly rebase this series to the latest master state?
> 
> Thanks!

@xback thanks for the heads-up! I rebased, tested, and pushed.

**Comment by xback** [comment]
@Gateworks
Thanks for the update.

I will give this series another spin based on kernel 6.1 tomorrow morning on a venice board.

**Comment by xback** [comment]
@Gateworks 

I tested the latest state a few moments ago as promised.
Using kernel 6.1 indeed yields a lot better results!

For reference, here is the [full bootlog](https://pastebin.com/raw/ZHQpjpLR) with some additional details at the bottom.

A few details left:

1)
- Will you also switch WAN & LAN ports as you did for the Ventana boards?  this ways these boards follow the same network port logic

2)
The boot log shows following line:

[   10.670041] I/O error, dev loop0, sector 12282 op 0x9:(WRITE_ZEROES) flags 0x800 phys_seg 0 prio class 2

Any idea?


For the rest .. all works nicely.



I also tested a mix of a ath9k radio with a ath11k radio (IRQ vs MSI) which didn't work well on a ventana IIRC.
Looks like this works very nicely here:

```
root@OpenWrt:/# lspci
00:00.0 PCI bridge: Synopsys, Inc. DWC_usb3 / PCIe bridge (rev 01)
...
03:00.0 Network controller: Qualcomm Technologies, Inc QCN6024/9024/9074 Wireless Network Adapter (rev 01)
04:00.0 Network controller: Qualcomm Atheros AR958x 802.11abgn Wireless Network Adapter (rev 01)
...
```

**Review by xback — APPROVED** [review]
Fully tested on an actual Venice board.
All works fine.

**Comment by Gateworks** [comment]
> @Gateworks
> 
> I tested the latest state a few moments ago as promised. Using kernel 6.1 indeed yields a lot better results!
> 
> For reference, here is the [full bootlog](https://pastebin.com/raw/ZHQpjpLR) with some additional details at the bottom.
> 

@xback 

Thank you for testing!

> A few details left:
> 
> 
> Will you also switch WAN & LAN ports as you did for the Ventana boards? this way these boards follow the same network port logic
> 

This should already be in place via the 2nd patch in the series that adds configuration for the gw72xx/gw73xx boards. I do still need to add the network config for the imx8mp gw74xx but PCI, USB, EQOS (2nd network port) require some backports for that (which I have queued up but I didn't want to add here as I didn't want to slow this series down)

> 
> The boot log shows following line: [ 10.670041] I/O error, dev loop0, sector 12282 op 0x9:(WRITE_ZEROES) flags 0x800 phys_seg 0 prio class 2

I don't see this... some overlay issue? Is it repeatable?

I'm installing bin/targets/imx/cortexa53/openwrt-imx-cortexa53-gateworks_venice-squashfs-img.gz via U-Boot via the 'update_all' script (tftpboot $loadaddr openwrt-imx-cortexa53-gateworks_venice-squashfs-img.gz && gzwrite mmc $dev $loadaddr $filesize)

root@OpenWrt:/# dmesg | grep loop
[    0.012537] Calibrating delay loop (skipped), value calculated using timer frequency.. 16.00 BogoMIPS (lpj=80000)
[    0.533800] loop: module loaded
[    4.558613] loop0: detected capacity change from 0 to 212992
[    4.621536] loop0: detected capacity change from 212992 to 207232
[    4.628950] mount_root: overlay filesystem in /dev/loop0 has not been formatted yet
[    4.744200] F2FS-fs (loop0): Found nat_bits in checkpoint
[    4.756148] F2FS-fs (loop0): Mounted with checkpoint version = 47aee43

I tested last with the bare minimum config:
~/openwrt$ ./scripts/diffconfig.sh 
CONFIG_TARGET_imx=y
CONFIG_TARGET_imx_cortexa53=y
CONFIG_TARGET_imx_cortexa53_Default=y
CONFIG_ALL_KMODS=y
CONFIG_BUILD_NLS=y
CONFIG_KERNEL_WERROR=y
CONFIG_TARGET_INITRAMFS_COMPRESSION_NONE=y
CONFIG_TARGET_ROOTFS_INITRAMFS=y

> 
> and this one [ 0.454568] imx-gpcv2 303a0000.gpc: Failed to create device link (0x180) with 30390000.reset-controller
> 

This one looks like it came in with 6.1.16. I don't know what it is caused by but I don't see an issue with it. The reset-controller is used for PCI which appears to be working fine.

**Comment by Gateworks** [comment]
@xback I just force pushed a refresh that has the following:
- added GW74xx network default config
- added some defines to cortexa53/config-default required for imx8mp USB/EQOS
- added some backported patches required for imx8mp PCI
- refresh the cortexta53/config-default (sorted properly)

**Comment by Gateworks** [comment]
@pepe2k @Ansuel could you please review this?

**Comment by robimarko** [comment]
Thanks! Rebased on top of main and merged!

**Comment by robimarko** [comment]
Its been open for long enough without any negative comments :)
